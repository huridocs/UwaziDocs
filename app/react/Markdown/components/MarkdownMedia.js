import PropTypes from 'prop-types';
import React, { Component } from 'react';
import ReactPlayer from 'react-player';

export class MarkdownMedia extends Component {
  propsToConfig(props) {
    const config = {
      url: '',
      options: {}
    };

    let parsedProps = props.config.replace(/\(|\)/g, '').split(',');
    config.url = parsedProps.shift();

    parsedProps = parsedProps.join(',') || '{}';
    parsedProps = parsedProps.replace(/&quot;/g, '"');

    try {
      parsedProps = JSON.parse(parsedProps);
    } catch (error) {
      parsedProps = {};
      console.log(error);
    }

    config.options = parsedProps;

    return config;
  }

  timeLinks(timelinks) {
    timelinks = timelinks || {};
    return Object.keys(timelinks).map((timeKey, index) => {
      const seconds = timeKey.split(':').reverse().reduce((_seconds, n, _index) => _seconds + parseInt(n, 10) * (60 ** _index), 0);
      return (<div className="timelink" key={index} onClick={this.seekTo.bind(this, seconds)} >
        <b><i className="fa fa-play" /> {timeKey}</b>
        <span>{timelinks[timeKey]}</span>
      </div>);
    });
  }

  seekTo(seconds) {
    this.refs.player.seekTo(seconds);
  }

  render() {
    const config = this.propsToConfig(this.props);

    return (
      <div className="video-container">
        <div>
          <ReactPlayer
            ref="player"
            width="100%"
            height="100%"
            url={config.url}
            controls
          />
        </div>
        <div>
          {this.timeLinks(config.options.timelinks)}
        </div>
      </div>
    );
  }
}

MarkdownMedia.propTypes = {
  config: PropTypes.string
};

export default MarkdownMedia;
