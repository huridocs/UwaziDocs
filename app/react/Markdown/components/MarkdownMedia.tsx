/* eslint-disable max-lines */
/* eslint-disable max-statements */
import { Translate } from 'app/I18N';
import React, { createRef, LegacyRef, useState } from 'react';
import { FieldArrayWithId, useFieldArray, useForm } from 'react-hook-form';
import ReactPlayer from 'react-player';
import { Icon } from 'UI';

export interface MarkdownMediaProps {
  compact?: boolean;
  editing?: boolean;
  onTimeLinkAdded?: Function;
  config: string;
}

interface TimeLink {
  timeHours: string;
  timeMinutes: string;
  timeSeconds: string;
  label: string;
}

const propsToConfig = (props: MarkdownMediaProps) => {
  const config = { url: '', options: { timelinks: {} } };

  let parsedProps: any = props.config.replace(/\(|\)/g, '').split(',');
  config.url = parsedProps.shift() as string;

  parsedProps = (parsedProps.join(',') || '{}').replace(/&quot;/g, '"');

  try {
    parsedProps = JSON.parse(parsedProps);
  } catch (error) {
    parsedProps = {};
  }

  config.options = parsedProps;

  return config;
};

const formatTimeLinks = (timelinks: any): TimeLink[] =>
  Object.keys(timelinks).map(key => {
    const timeLink = { timeHours: '00', timeMinutes: '00', timeSeconds: '00', label: '' };
    const splitTimes = key.split(':');
    if (splitTimes.length === 2) {
      [timeLink.timeMinutes, timeLink.timeSeconds] = splitTimes;
    } else {
      [timeLink.timeHours, timeLink.timeMinutes, timeLink.timeSeconds] = splitTimes;
    }
    timeLink.label = timelinks[key] as string;
    return timeLink;
  });

const MarkdownMedia = (props: MarkdownMediaProps) => {
  const [newTimeline, setNewTimeline] = useState<TimeLink>({
    timeHours: '00',
    timeMinutes: '00',
    timeSeconds: '00',
    label: '',
  });
  const { options } = propsToConfig(props);
  const originalTimelinks = formatTimeLinks(options?.timelinks || {});
  const [playingTimelinkIndex, setPlayingTimelinkIndex] = useState<number>(-1);
  const [isVideoPlaying, setVideoPlaying] = useState<boolean>(false);

  const { control, register, getValues } = useForm<{ timelines: TimeLink[] }>({
    defaultValues: { timelines: originalTimelinks },
  });
  const { fields, remove, append } = useFieldArray<{ timelines: TimeLink[] }>({
    control,
    name: 'timelines',
  });

  const playerRef: LegacyRef<ReactPlayer> | undefined = createRef();

  const seekTo = (seconds: number) => {
    const playingStatus = isVideoPlaying;
    playerRef.current?.seekTo(seconds);
    setVideoPlaying(playingStatus);
  };

  const constructTimelinksString = (timelinks: TimeLink[], url: string) => {
    const timelinksObj = timelinks.reduce((current: any, timelink) => {
      current[`${timelink.timeHours}:${timelink.timeMinutes}:${timelink.timeSeconds}`] =
        timelink.label;
      return current;
    }, {});
    return `(${url}, ${JSON.stringify({ timelinks: timelinksObj })})`;
  };

  const timeLinks = (_timelinks: any) => {
    const timelinks = _timelinks || {};
    return Object.keys(timelinks).map((timeKey, index) => {
      const seconds = timeKey
        .split(':')
        .reverse()
        .reduce((_seconds, n, _index) => _seconds + parseInt(n, 10) * 60 ** _index, 0);

      let displayTime = timeKey;
      if (displayTime.split(':').length === 2) {
        displayTime = `00:${displayTime}`;
      }

      return (
        <div className="timelink" key={timeKey + index.toString()} title={timelinks[timeKey]}>
          <b
            className="timelink-icon"
            onClick={() => {
              seekTo(seconds);
              if (index === playingTimelinkIndex) {
                setVideoPlaying(false);
                setPlayingTimelinkIndex(-1);
              } else {
                setVideoPlaying(true);
                setPlayingTimelinkIndex(index);
              }
            }}
          >
            <Icon icon={index === playingTimelinkIndex ? 'pause' : 'play'} />
          </b>
          <div
            className="timelink-time-label"
            onClick={() => {
              seekTo(seconds);
              setPlayingTimelinkIndex(index);
            }}
          >
            <b>{displayTime}</b>
            <span>{timelinks[timeKey]}</span>
          </div>
        </div>
      );
    });
  };

  const updateParentForm = (url: string) => {
    const fullTimelinksString = constructTimelinksString(getValues().timelines, url);
    if (props.onTimeLinkAdded) props.onTimeLinkAdded(fullTimelinksString);
  };

  const appendTimelinkAndUpdateParent = (url: string, timelink?: TimeLink) => {
    const currentTimelink = timelink || newTimeline;
    append(currentTimelink);
    updateParentForm(url);
  };

  const renderNewTimeLinkForm = (url: string) => (
    <div className="new-timelink">
      <div className="timestamp">
        <input
          type="text"
          onChange={event => {
            setNewTimeline({ ...newTimeline, timeHours: event.target.value });
          }}
          className="timestamp-hours"
          placeholder="00"
          required
          value={newTimeline.timeHours}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          onChange={event => {
            setNewTimeline({ ...newTimeline, timeMinutes: event.target.value });
          }}
          className="timestamp-minutes"
          placeholder="00"
          required
          value={newTimeline.timeMinutes}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          onChange={event => {
            setNewTimeline({ ...newTimeline, timeSeconds: event.target.value });
          }}
          className="timestamp-seconds"
          placeholder="00"
          required
          value={newTimeline.timeSeconds}
        />
      </div>
      <input
        type="text"
        onChange={event => {
          setNewTimeline({ ...newTimeline, label: event.target.value });
        }}
        className="timestamp-label"
        placeholder="Enter title"
        value={newTimeline.label}
      />
      <button
        type="button"
        className="new-timestamp-btn"
        onClick={() => {
          appendTimelinkAndUpdateParent(url);
          setNewTimeline({ timeHours: '00', timeMinutes: '00', timeSeconds: '00', label: '' });
        }}
      >
        <Icon icon="plus" />
      </button>
    </div>
  );

  const renderSingleTimeLinkInputs = (
    field: FieldArrayWithId<TimeLink>,
    url: string,
    index: number
  ) => (
    <div className="new-timelink" key={index}>
      <div className="timestamp">
        <input
          type="text"
          className="timestamp-hours"
          placeholder="00"
          key={field.id}
          {...register(`timelines.${index}.timeHours`, {
            onChange: _ => updateParentForm(url),
          })}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          className="timestamp-minutes"
          placeholder="00"
          key={field.id}
          {...register(`timelines.${index}.timeMinutes`, {
            onChange: _ => updateParentForm(url),
          })}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          className="timestamp-seconds"
          placeholder="00"
          key={field.id}
          {...register(`timelines.${index}.timeSeconds`, {
            onChange: _ => updateParentForm(url),
          })}
        />
      </div>
      <input
        type="text"
        className="timestamp-label"
        placeholder="Enter title"
        key={field.id}
        {...register(`timelines.${index}.label`, {
          onChange: _ => updateParentForm(url),
        })}
      />
      <button
        title="Remove timelink"
        type="button"
        className="delete-timestamp-btn"
        onClick={() => {
          remove(index);
          updateParentForm(url);
        }}
      >
        <Icon icon="trash-alt" />
      </button>
    </div>
  );

  const renderTimeLinksForm = (url: string) => (
    <>
      {fields.map((field: FieldArrayWithId<TimeLink>, index: number) =>
        renderSingleTimeLinkInputs(field, url, index)
      )}
      {renderNewTimeLinkForm(url)}
    </>
  );

  const config = propsToConfig(props);
  const { compact, editing } = props;
  const dimensions: { width: string; height?: string } = { width: '100%' };
  if (compact) {
    dimensions.height = '100%';
  }

  return (
    <div className={`video-container ${compact ? 'compact' : ''}`}>
      <div>
        <ReactPlayer
          className="react-player"
          playing={isVideoPlaying}
          ref={playerRef}
          url={config.url}
          {...dimensions}
          controls
          onPause={() => {
            setVideoPlaying(false);
          }}
          onPlay={() => {
            setVideoPlaying(true);
          }}
        />
      </div>
      {!editing && <div>{timeLinks(config.options.timelinks)}</div>}
      {editing && (
        <div className="timelinks-form">
          <button
            type="button"
            className="add-timelink"
            onClick={() => {
              const currentTime = playerRef.current?.getCurrentTime() as number;
              const hours = Math.floor(currentTime / 3600);
              const remainingSeconds = currentTime - hours * 3600;
              const minutes = Math.floor(remainingSeconds / 60);
              const seconds = Math.floor(remainingSeconds % 60);
              const timelink = {
                timeHours: hours < 10 ? `0${hours.toString()}` : hours.toString(),
                timeMinutes: minutes < 10 ? `0${minutes.toString()}` : minutes.toString(),
                timeSeconds: seconds < 10 ? `0${seconds.toString()}` : seconds.toString(),
                label: '',
              };
              appendTimelinkAndUpdateParent(config.url, timelink);
            }}
          >
            <Translate>Add timelink</Translate>
          </button>
          <h5>
            <Translate>Timelinks</Translate>
          </h5>
          {renderTimeLinksForm(config.url)}
        </div>
      )}
      <p className="print-view-alt">{config.url}</p>
    </div>
  );
};

export default MarkdownMedia;
