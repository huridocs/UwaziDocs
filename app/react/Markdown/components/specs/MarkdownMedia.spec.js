import React from 'react';
import MarkdownMedia from '../MarkdownMedia';
import {mount} from 'enzyme';

describe('MarkdownMedia', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      config: '(https://www.vimeo.com/253530307, {"timelinks": {"02:10": "A rude awakening", "05:30": "Finally, you are up!"}})'
    };
  });

  let render = () => {
    component = mount(<MarkdownMedia {...props} />);
  };

  describe('render', () => {
    it('should render an iframe with the correct video id', () => {
      render();
      expect(component.find('ReactPlayer').props().url)
      .toBe('https://www.vimeo.com/253530307');
    });


    describe('Video timeline', () => {
      it('should render the timelinks and call the player onClick', () => {
        render();

        const links = component.find('.timelink');
        expect(links.length).toBe(2);

        const player = component.instance().refs.player;
        spyOn(player, 'seekTo');
        const firstTimeLink = component.find('.timelink').at(0);
        const secondTimeLink = component.find('.timelink').at(1);
        firstTimeLink.simulate('click');
        expect(player.seekTo).toHaveBeenCalledWith(2 * 60 + 10);
        secondTimeLink.simulate('click');
        expect(player.seekTo).toHaveBeenCalledWith(5 * 60 + 30);
      });
    });
  });
});
