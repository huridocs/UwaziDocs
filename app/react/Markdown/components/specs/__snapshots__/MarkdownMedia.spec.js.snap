// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownMedia render should render an iframe with the correct video id 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component>
    <MarkdownMedia
      config="(https://www.vimeo.com/253530307, {\\"timelinks\\": {\\"02:10\\": \\"A rude awakening\\", \\"05:30\\": \\"Finally, you are up!\\"}})"
    >
      <div
        className="video-container "
      >
        <div>
          <ReactPlayer
            className="react-player"
            config={
              Object {
                "dailymotion": Object {
                  "params": Object {
                    "api": 1,
                    "endscreen-enable": false,
                  },
                },
                "facebook": Object {
                  "appId": "1309697205772819",
                  "attributes": Object {},
                  "playerId": null,
                  "version": "v3.3",
                },
                "file": Object {
                  "attributes": Object {},
                  "dashVersion": "3.1.3",
                  "flvVersion": "1.5.0",
                  "forceAudio": false,
                  "forceDASH": false,
                  "forceFLV": false,
                  "forceHLS": false,
                  "forceVideo": false,
                  "hlsOptions": Object {},
                  "hlsVersion": "1.1.4",
                  "tracks": Array [],
                },
                "mixcloud": Object {
                  "options": Object {
                    "hide_cover": 1,
                  },
                },
                "soundcloud": Object {
                  "options": Object {
                    "buying": false,
                    "download": false,
                    "liking": false,
                    "sharing": false,
                    "show_comments": false,
                    "show_playcount": false,
                    "visual": true,
                  },
                },
                "twitch": Object {
                  "options": Object {},
                  "playerId": null,
                },
                "vidyard": Object {
                  "options": Object {},
                },
                "vimeo": Object {
                  "playerOptions": Object {
                    "autopause": false,
                    "byline": false,
                    "portrait": false,
                    "title": false,
                  },
                  "title": null,
                },
                "wistia": Object {
                  "customControls": null,
                  "options": Object {},
                  "playerId": null,
                },
                "youtube": Object {
                  "embedOptions": Object {},
                  "onUnstarted": [Function],
                  "playerVars": Object {
                    "iv_load_policy": 3,
                    "modestbranding": 1,
                    "playsinline": 1,
                    "rel": 0,
                    "showinfo": 0,
                  },
                },
              }
            }
            controls={true}
            fallback={null}
            height="360px"
            light={false}
            loop={false}
            muted={false}
            oEmbedUrl="https://noembed.com/embed?url={url}"
            onBuffer={[Function]}
            onBufferEnd={[Function]}
            onClickPreview={[Function]}
            onDisablePIP={[Function]}
            onDuration={[Function]}
            onEnablePIP={[Function]}
            onEnded={[Function]}
            onError={[Function]}
            onPause={[Function]}
            onPlay={[Function]}
            onPlaybackRateChange={[Function]}
            onProgress={[Function]}
            onReady={[Function]}
            onSeek={[Function]}
            onStart={[Function]}
            pip={false}
            playbackRate={1}
            playing={false}
            playsinline={false}
            previewTabIndex={0}
            progressInterval={1000}
            stopOnUnmount={true}
            style={Object {}}
            url="https://www.vimeo.com/253530307"
            volume={null}
            width="100%"
            wrapper="div"
          >
            <div
              className="react-player"
              style={
                Object {
                  "height": "360px",
                  "width": "100%",
                }
              }
            >
              <Suspense
                fallback={null}
              />
            </div>
          </ReactPlayer>
        </div>
        <div>
          <div
            className="timelink"
            key="02:100"
            title="A rude awakening"
          >
            <b
              className="timelink-icon"
              onClick={[Function]}
            >
              <Connect(Icon)
                icon="play"
              >
                <Icon
                  directionAware={false}
                  icon="play"
                  locale=""
                >
                  <FontAwesomeIcon
                    beat={false}
                    beatFade={false}
                    border={false}
                    bounce={false}
                    className=""
                    fade={false}
                    fixedWidth={false}
                    flip={null}
                    icon="play"
                    inverse={false}
                    listItem={false}
                    mask={null}
                    maskId={null}
                    pull={null}
                    pulse={false}
                    rotation={null}
                    shake={false}
                    size={null}
                    spin={false}
                    spinPulse={false}
                    spinReverse={false}
                    swapOpacity={false}
                    symbol={false}
                    title=""
                    titleId={null}
                    transform={null}
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-play fa-w-14 "
                      data-icon="play"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </FontAwesomeIcon>
                </Icon>
              </Connect(Icon)>
            </b>
            <div
              className="timelink-time-label"
              onClick={[Function]}
            >
              <b>
                00:02:10
              </b>
              <span>
                A rude awakening
              </span>
            </div>
          </div>
          <div
            className="timelink"
            key="05:301"
            title="Finally, you are up!"
          >
            <b
              className="timelink-icon"
              onClick={[Function]}
            >
              <Connect(Icon)
                icon="play"
              >
                <Icon
                  directionAware={false}
                  icon="play"
                  locale=""
                >
                  <FontAwesomeIcon
                    beat={false}
                    beatFade={false}
                    border={false}
                    bounce={false}
                    className=""
                    fade={false}
                    fixedWidth={false}
                    flip={null}
                    icon="play"
                    inverse={false}
                    listItem={false}
                    mask={null}
                    maskId={null}
                    pull={null}
                    pulse={false}
                    rotation={null}
                    shake={false}
                    size={null}
                    spin={false}
                    spinPulse={false}
                    spinReverse={false}
                    swapOpacity={false}
                    symbol={false}
                    title=""
                    titleId={null}
                    transform={null}
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-play fa-w-14 "
                      data-icon="play"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </FontAwesomeIcon>
                </Icon>
              </Connect(Icon)>
            </b>
            <div
              className="timelink-time-label"
              onClick={[Function]}
            >
              <b>
                00:05:30
              </b>
              <span>
                Finally, you are up!
              </span>
            </div>
          </div>
        </div>
        <p
          className="print-view-alt"
        >
          https://www.vimeo.com/253530307
        </p>
      </div>
    </MarkdownMedia>
  </Component>
</Provider>
`;

exports[`MarkdownMedia render should use compact class name if compact prop is set 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component>
    <MarkdownMedia
      compact={true}
      config="(https://www.vimeo.com/253530307, {\\"timelinks\\": {\\"02:10\\": \\"A rude awakening\\", \\"05:30\\": \\"Finally, you are up!\\"}})"
    >
      <div
        className="video-container compact"
      >
        <div>
          <ReactPlayer
            className="react-player"
            config={
              Object {
                "dailymotion": Object {
                  "params": Object {
                    "api": 1,
                    "endscreen-enable": false,
                  },
                },
                "facebook": Object {
                  "appId": "1309697205772819",
                  "attributes": Object {},
                  "playerId": null,
                  "version": "v3.3",
                },
                "file": Object {
                  "attributes": Object {},
                  "dashVersion": "3.1.3",
                  "flvVersion": "1.5.0",
                  "forceAudio": false,
                  "forceDASH": false,
                  "forceFLV": false,
                  "forceHLS": false,
                  "forceVideo": false,
                  "hlsOptions": Object {},
                  "hlsVersion": "1.1.4",
                  "tracks": Array [],
                },
                "mixcloud": Object {
                  "options": Object {
                    "hide_cover": 1,
                  },
                },
                "soundcloud": Object {
                  "options": Object {
                    "buying": false,
                    "download": false,
                    "liking": false,
                    "sharing": false,
                    "show_comments": false,
                    "show_playcount": false,
                    "visual": true,
                  },
                },
                "twitch": Object {
                  "options": Object {},
                  "playerId": null,
                },
                "vidyard": Object {
                  "options": Object {},
                },
                "vimeo": Object {
                  "playerOptions": Object {
                    "autopause": false,
                    "byline": false,
                    "portrait": false,
                    "title": false,
                  },
                  "title": null,
                },
                "wistia": Object {
                  "customControls": null,
                  "options": Object {},
                  "playerId": null,
                },
                "youtube": Object {
                  "embedOptions": Object {},
                  "onUnstarted": [Function],
                  "playerVars": Object {
                    "iv_load_policy": 3,
                    "modestbranding": 1,
                    "playsinline": 1,
                    "rel": 0,
                    "showinfo": 0,
                  },
                },
              }
            }
            controls={true}
            fallback={null}
            height="100%"
            light={false}
            loop={false}
            muted={false}
            oEmbedUrl="https://noembed.com/embed?url={url}"
            onBuffer={[Function]}
            onBufferEnd={[Function]}
            onClickPreview={[Function]}
            onDisablePIP={[Function]}
            onDuration={[Function]}
            onEnablePIP={[Function]}
            onEnded={[Function]}
            onError={[Function]}
            onPause={[Function]}
            onPlay={[Function]}
            onPlaybackRateChange={[Function]}
            onProgress={[Function]}
            onReady={[Function]}
            onSeek={[Function]}
            onStart={[Function]}
            pip={false}
            playbackRate={1}
            playing={false}
            playsinline={false}
            previewTabIndex={0}
            progressInterval={1000}
            stopOnUnmount={true}
            style={Object {}}
            url="https://www.vimeo.com/253530307"
            volume={null}
            width="100%"
            wrapper="div"
          >
            <div
              className="react-player"
              style={
                Object {
                  "height": "100%",
                  "width": "100%",
                }
              }
            >
              <Suspense
                fallback={null}
              >
                <Player
                  activePlayer={
                    Object {
                      "$$typeof": Symbol(react.lazy),
                      "_init": [Function],
                      "_payload": Object {
                        "_result": Object {
                          "default": [Function],
                        },
                        "_status": 1,
                      },
                    }
                  }
                  className="react-player"
                  config={
                    Object {
                      "dailymotion": Object {
                        "params": Object {
                          "api": 1,
                          "endscreen-enable": false,
                        },
                      },
                      "facebook": Object {
                        "appId": "1309697205772819",
                        "attributes": Object {},
                        "playerId": null,
                        "version": "v3.3",
                      },
                      "file": Object {
                        "attributes": Object {},
                        "dashVersion": "3.1.3",
                        "flvVersion": "1.5.0",
                        "forceAudio": false,
                        "forceDASH": false,
                        "forceFLV": false,
                        "forceHLS": false,
                        "forceVideo": false,
                        "hlsOptions": Object {},
                        "hlsVersion": "1.1.4",
                        "tracks": Array [],
                      },
                      "mixcloud": Object {
                        "options": Object {
                          "hide_cover": 1,
                        },
                      },
                      "playerOptions": Object {
                        "autopause": false,
                        "byline": false,
                        "portrait": false,
                        "title": false,
                      },
                      "soundcloud": Object {
                        "options": Object {
                          "buying": false,
                          "download": false,
                          "liking": false,
                          "sharing": false,
                          "show_comments": false,
                          "show_playcount": false,
                          "visual": true,
                        },
                      },
                      "title": null,
                      "twitch": Object {
                        "options": Object {},
                        "playerId": null,
                      },
                      "vidyard": Object {
                        "options": Object {},
                      },
                      "vimeo": Object {
                        "playerOptions": Object {
                          "autopause": false,
                          "byline": false,
                          "portrait": false,
                          "title": false,
                        },
                        "title": null,
                      },
                      "wistia": Object {
                        "customControls": null,
                        "options": Object {},
                        "playerId": null,
                      },
                      "youtube": Object {
                        "embedOptions": Object {},
                        "onUnstarted": [Function],
                        "playerVars": Object {
                          "iv_load_policy": 3,
                          "modestbranding": 1,
                          "playsinline": 1,
                          "rel": 0,
                          "showinfo": 0,
                        },
                      },
                    }
                  }
                  controls={true}
                  fallback={null}
                  height="100%"
                  key="vimeo"
                  light={false}
                  loop={false}
                  muted={false}
                  oEmbedUrl="https://noembed.com/embed?url={url}"
                  onBuffer={[Function]}
                  onBufferEnd={[Function]}
                  onClickPreview={[Function]}
                  onDisablePIP={[Function]}
                  onDuration={[Function]}
                  onEnablePIP={[Function]}
                  onEnded={[Function]}
                  onError={[Function]}
                  onPause={[Function]}
                  onPlay={[Function]}
                  onPlaybackRateChange={[Function]}
                  onProgress={[Function]}
                  onReady={[Function]}
                  onSeek={[Function]}
                  onStart={[Function]}
                  pip={false}
                  playbackRate={1}
                  playing={false}
                  playsinline={false}
                  previewTabIndex={0}
                  progressInterval={1000}
                  stopOnUnmount={true}
                  style={Object {}}
                  url="https://www.vimeo.com/253530307"
                  volume={null}
                  width="100%"
                  wrapper="div"
                >
                  <Vimeo
                    activePlayer={
                      Object {
                        "$$typeof": Symbol(react.lazy),
                        "_init": [Function],
                        "_payload": Object {
                          "_result": Object {
                            "default": [Function],
                          },
                          "_status": 1,
                        },
                      }
                    }
                    className="react-player"
                    config={
                      Object {
                        "dailymotion": Object {
                          "params": Object {
                            "api": 1,
                            "endscreen-enable": false,
                          },
                        },
                        "facebook": Object {
                          "appId": "1309697205772819",
                          "attributes": Object {},
                          "playerId": null,
                          "version": "v3.3",
                        },
                        "file": Object {
                          "attributes": Object {},
                          "dashVersion": "3.1.3",
                          "flvVersion": "1.5.0",
                          "forceAudio": false,
                          "forceDASH": false,
                          "forceFLV": false,
                          "forceHLS": false,
                          "forceVideo": false,
                          "hlsOptions": Object {},
                          "hlsVersion": "1.1.4",
                          "tracks": Array [],
                        },
                        "mixcloud": Object {
                          "options": Object {
                            "hide_cover": 1,
                          },
                        },
                        "playerOptions": Object {
                          "autopause": false,
                          "byline": false,
                          "portrait": false,
                          "title": false,
                        },
                        "soundcloud": Object {
                          "options": Object {
                            "buying": false,
                            "download": false,
                            "liking": false,
                            "sharing": false,
                            "show_comments": false,
                            "show_playcount": false,
                            "visual": true,
                          },
                        },
                        "title": null,
                        "twitch": Object {
                          "options": Object {},
                          "playerId": null,
                        },
                        "vidyard": Object {
                          "options": Object {},
                        },
                        "vimeo": Object {
                          "playerOptions": Object {
                            "autopause": false,
                            "byline": false,
                            "portrait": false,
                            "title": false,
                          },
                          "title": null,
                        },
                        "wistia": Object {
                          "customControls": null,
                          "options": Object {},
                          "playerId": null,
                        },
                        "youtube": Object {
                          "embedOptions": Object {},
                          "onUnstarted": [Function],
                          "playerVars": Object {
                            "iv_load_policy": 3,
                            "modestbranding": 1,
                            "playsinline": 1,
                            "rel": 0,
                            "showinfo": 0,
                          },
                        },
                      }
                    }
                    controls={true}
                    fallback={null}
                    height="100%"
                    light={false}
                    loop={false}
                    muted={false}
                    oEmbedUrl="https://noembed.com/embed?url={url}"
                    onBuffer={[Function]}
                    onBufferEnd={[Function]}
                    onClickPreview={[Function]}
                    onDisablePIP={[Function]}
                    onDuration={[Function]}
                    onEnablePIP={[Function]}
                    onEnded={[Function]}
                    onError={[Function]}
                    onLoaded={[Function]}
                    onMount={[Function]}
                    onPause={[Function]}
                    onPlay={[Function]}
                    onPlaybackRateChange={[Function]}
                    onProgress={[Function]}
                    onReady={[Function]}
                    onSeek={[Function]}
                    onStart={[Function]}
                    pip={false}
                    playbackRate={1}
                    playing={false}
                    playsinline={false}
                    previewTabIndex={0}
                    progressInterval={1000}
                    stopOnUnmount={true}
                    style={Object {}}
                    url="https://www.vimeo.com/253530307"
                    volume={null}
                    width="100%"
                    wrapper="div"
                  >
                    <div
                      key="https://www.vimeo.com/253530307"
                      style={
                        Object {
                          "display": undefined,
                          "height": "100%",
                          "overflow": "hidden",
                          "width": "100%",
                        }
                      }
                    />
                  </Vimeo>
                </Player>
              </Suspense>
            </div>
          </ReactPlayer>
        </div>
        <div>
          <div
            className="timelink"
            key="02:100"
            title="A rude awakening"
          >
            <b
              className="timelink-icon"
              onClick={[Function]}
            >
              <Connect(Icon)
                icon="play"
              >
                <Icon
                  directionAware={false}
                  icon="play"
                  locale=""
                >
                  <FontAwesomeIcon
                    beat={false}
                    beatFade={false}
                    border={false}
                    bounce={false}
                    className=""
                    fade={false}
                    fixedWidth={false}
                    flip={null}
                    icon="play"
                    inverse={false}
                    listItem={false}
                    mask={null}
                    maskId={null}
                    pull={null}
                    pulse={false}
                    rotation={null}
                    shake={false}
                    size={null}
                    spin={false}
                    spinPulse={false}
                    spinReverse={false}
                    swapOpacity={false}
                    symbol={false}
                    title=""
                    titleId={null}
                    transform={null}
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-play fa-w-14 "
                      data-icon="play"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </FontAwesomeIcon>
                </Icon>
              </Connect(Icon)>
            </b>
            <div
              className="timelink-time-label"
              onClick={[Function]}
            >
              <b>
                00:02:10
              </b>
              <span>
                A rude awakening
              </span>
            </div>
          </div>
          <div
            className="timelink"
            key="05:301"
            title="Finally, you are up!"
          >
            <b
              className="timelink-icon"
              onClick={[Function]}
            >
              <Connect(Icon)
                icon="play"
              >
                <Icon
                  directionAware={false}
                  icon="play"
                  locale=""
                >
                  <FontAwesomeIcon
                    beat={false}
                    beatFade={false}
                    border={false}
                    bounce={false}
                    className=""
                    fade={false}
                    fixedWidth={false}
                    flip={null}
                    icon="play"
                    inverse={false}
                    listItem={false}
                    mask={null}
                    maskId={null}
                    pull={null}
                    pulse={false}
                    rotation={null}
                    shake={false}
                    size={null}
                    spin={false}
                    spinPulse={false}
                    spinReverse={false}
                    swapOpacity={false}
                    symbol={false}
                    title=""
                    titleId={null}
                    transform={null}
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-play fa-w-14 "
                      data-icon="play"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </FontAwesomeIcon>
                </Icon>
              </Connect(Icon)>
            </b>
            <div
              className="timelink-time-label"
              onClick={[Function]}
            >
              <b>
                00:05:30
              </b>
              <span>
                Finally, you are up!
              </span>
            </div>
          </div>
        </div>
        <p
          className="print-view-alt"
        >
          https://www.vimeo.com/253530307
        </p>
      </div>
    </MarkdownMedia>
  </Component>
</Provider>
`;
